### ISSUE TITLE
Bind transcoding options with transcoded results in orchestrator signature

### ISSUE THREAD

**Issue by yondonfu** [livepeer/go-livepeer#1091]
**Is your feature request related to a problem? Please describe.**

At the moment, an orchestrator in on-chain mode will [sign the following payload](https://github.com/livepeer/go-livepeer/blob/master/core/orchestrator.go#L546):

```
H(H(res_1) + H(res_2) + ... + H(res_n))
```

where `res_1, res_2, ..., res_n` is the set of transcoded results for a source segment. The signature over this hash can be used by the broadcaster to authenticate the set of transcoded results received from an orchestrator. However, the signature does not bind a specific set of transcoding options with each result. 

**Describe the solution you'd like**

Bind a set of transcoding options with each result with the orchestrator's signature. We can change the payload signed by the orchestrator to:

```
H(H(res_1 + t_opts_1) + H(res_2 + t_opts_2) + ... + H(res_n + t_opts_n))
```

where `t_opts_1, t_opts_2, ..., t_opts_n` is the set of transcoding options associated with the set of transcoded results where `t_opts_i` corresponds to `res_i`.

Each `t_opt_i` can be the [4 byte identifier](https://github.com/livepeer/go-livepeer/blob/master/common/videoprofile_ids.go) associated with the video profile associated with the result. 

This will require:

- [ ] Updating the orchestrator signing logic
- [ ] Updating the broadcaster's verification logic

**Comment by j0sh** [comment]
> Each t_opt_i can be the 4 byte identifier

I'm unsure how those 4 byte identifiers were generated; is it the first bytes of some hashed construction of the VideoProfile?

At some point we'll need to expand beyond our current fixed set of video profiles, and into arbitrary resolutions (at a minimum).

If we standardize on how to incorporate the [VideoProfile parameters](https://github.com/livepeer/lpms/blob/0246c39344c63ce2f2cd0bbc17d6704d35c998bc/ffmpeg/videoprofile.go#L20-L26) into the signature, then that can be used.

I am a little nervous about forward compatibility because it's possible (likely) that we'll add encoding options that aren't currently reflected in the VideoProfile, and a naive application of this scheme would force B/O/V to upgrade in lockstep. I can see arguments for/against a looser approach towards compatibility; let me think about it a little more.

**Comment by yondonfu** [comment]
> I'm unsure how those 4 byte identifiers were generated; is it the first bytes of some hashed construction of the VideoProfile?

An identifier consists of the first 4 bytes of the keccak256 hash of the video profile name

Ex. `"P720p60fps16x9" -> keccak256("P720p60fps16x9")[0:4] = a7ac137a`

The identifier format was defined for V1 when a list of video profile IDs needed to be submitted on-chain. Not wedded to this scheme though and open to other approaches.

**Comment by j0sh** [comment]
> I am a little nervous about forward compatibility because it's possible (likely) that we'll add encoding options that aren't currently reflected in the VideoProfile and a naive application of this scheme would force B/O/V to upgrade in lockstep

To mitigate this issue, how about the following scheme for generating the hash of each profile:

* Define a `net.VideoProfile` protocol buffers struct (we have to do this anyway for RPC)
* Serialize the struct (probably using protobuf's default serialization?)
* Take the hash of the serialization

Serializations such as JSON are possible [1], but the ordering of fields in JSON is undefined (so we'd have to do that ourselves), while they are defined for protobufs.

[1] and protobufs has a JSON mode but I'm still not sure if we should be using it for this hash

**Comment by yondonfu** [comment]
Relying on protobuf serialization seems reasonable, but we should just be mindful about the possibility of non-determinism [1].

If we update `net.SegData` to include an array of `net.VideoProfile` (instead of the current `bytes profiles`) we could serialize the `net.SegData` struct and hash that. `sig` is currently a field in `net.SegData` so that could just be omitted when serializing/hashing during signature generation/verification.

[1] This [gist](https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510) describes different reasons why non-deterministic serialization may occur. We should be safe from many of theses sources of non-determinism right now (i.e we don't use maps in our data structures). Unfortunately, while the protobuf implementation that we use will currently serialize/deserialize in a stable manner, the protobuf spec does not explicitly require it. That being said, I think the pro of relying on an existing serialization format instead of defining a custom one ultimately outweighs the con of potential non-determinism that we need to safeguard against if we update protobuf implementations and struct definitions in the future.

**Comment by j0sh** [comment]
Good point about protobuf serialization not necessarily being deterministic. I was thinking that the `net.VideoProfile` itself doesn't need to be re-serialized by the receiver; we only need to hash against the already-serialized data that we receive, and then validate that the decoded data is reasonable. But we can also make the exact same argument for JSON.

But this also feels a tiny bit like a trap, since we're shutting ourselves out of cases where we might still need to re-serialize and check.

> we could serialize the net.SegData struct and hash that. sig is currently a field in net.SegData so that could just be omitted when serializing/hashing during signature generation/verification.

That would also mean the following workflow when verifying the signature:
- Deserialize `net.SegData`
- Zero out the `sig` field
- Re-serialize `net.SegData`
- Hash and verify

> to include an array of net.VideoProfile (instead of the current bytes profiles) we could serialize the net.SegData

I had originally thought of using an serialized array of `net.VideoProfile` as part of the request hash, but wasn't sure how important it was to have the values for each profile tied to a specific output in the orchestrator's signature. If that's not important (and I suspect it's not), the two signatures could probably be generated like this:

```
// B request sig
SegData.Sig = Sign(H(H( []net.VideoProfile ) | H( source_Data )))

// O response sig
TranscodeData.Sig = Sign(H(H( SegData.Sig ) | H( res1_Data ) | ... | H(  resN_Data )))
```

Note there's no real need to incorporate the ManifestID or sequence number anymore, as far as I can tell.

**Comment by j0sh** [comment]
> the net.VideoProfile itself doesn't need to be re-serialized by the receiver; we only need to hash against the already-serialized data that we receive

Thinking about this for more than 30 seconds, it's going to be awkward validating this, because AFAIK it's not as easy to extract a single serialized field from a message, while still retaining the ability to deserialize the same field to the internal `net.VideoProfile` representation. A workaround would be to store `[]net.VideoProfile` in `SegData` as a byte array and explicitly manage that one field, but we give up on the safety and convenience of codegen.
